kotlin/samples/tictactoe/common/src/main/java/com/squareup/sample/panel/PanelContainerScreen.kt

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright 2019 Square Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.sample.panel
+
+import com.squareup.viewregistry.BackStackScreen
+import com.squareup.viewregistry.HasModals
+
+/**
+ * Custom modal container used in Tic Tac Workflow sample. Shows one or more
+ * nested [sub-flows][modals] over a [baseScreen].
+ */
+data class PanelContainerScreen<B : Any, T : Any>(
```

Comment:
Still curious about this name. It seems more related to stack structure than "panels", whatever those are.

Also, what is a Panel? Naming suggests it's some kind of first-class concept but I can't figure out what that is.

Reply:
It's just meant to mean _The Modal Thing Used in This Particular App to Shows Subflows_. In our app I'd have called it _Card_, but that has another meaning to Android devs in general.

_SubFlowContainerScreen_? _WizardContainerScreen_?

Reply:
`TicTacSubflow`?

Reply:
So a `Panel` is a… `Modal`? Does `Modal` mean something else?

Reply:
A modal is a thing in a dialog window. The most common form of a modal on Android is an alert dialog, so the library itself offers support for at least the basic form of those nasty pieces of shit.

The Tic Tac Toe application uses modals for two purposes:

 * Alerts, like everyone else
 * Panels: full screen (phone) or great big square (tablet) windows that host subtasks like logging in and choosing player names, screens that will likely start having nested toolbars before we're done but one thing at a time.

These "panels" are not general purpose, because unlike iOS's NavController they're not an Android-wide standard. That's why they're sample code. They're called panels because I had to call them something, and that was the first unused term that occurred to me.

In our own PoS app, we'll likely have two such things, Cards and Sheets, which look a lot like this sample but which enforce our own very Square-specific design language. 

Other apps by other companies may find AlertContainer to be all they need. Or they may fork it into something that can also show some of Android's stock pickers. Or they may also make something similar but not exactly like panel. But whatever they do will be app specific.

IMHO the abstract ModalContainer class has the base stick-some-shit-in-a-dialog support that is reasonable to maintain in a library. Any richer feature set has got to be app specific.

Does this help?

Reply:
> A modal is a thing in a dialog window. The most common form of a modal on Android is AlertDialog, so the library itself offers support for at least the basic form of those nasty pieces of shit. You'll find that React offers something astonishingly similar.
> 
> The Tic Tac Toe application uses modals for two purposes:
> 
> * Alerts, like everyone else
> * Panels: full screen (phone) or great big square (tablet) windows that host subtasks like logging in and choosing player names, tasks which take multiple steps and involve going backward and forward, using screens that will likely start having nested toolbars before we're done but one thing at a time.

Yes, this ☝️ cleared it up for me. I think I finally get it now 😅

I take it iOS doesn't have their own word for "panel"?

Reply:
`NavigationController`, basically. And for them, BackStackScreen is basically ThingThatShowsInNavigationController, unfortunately including info about the nested toolbar. 

Something to think about harder when we converge.

Reply:
```
/**
 * Custom modal container used in Tic Tac Workflow sample. Shows one or more
 * nested [sub-flows][modals] over a [baseScreen].
 *
 * Demonstrates how an app can set up a custom modal design element.
 * The Tic Tac Toe application uses modals for two purposes:
 * 
 *  - Alerts, like everyone else, via the stock `AlertContainerScreen`
 *  
 *  - Panels, this class: full screen (phone) or great big square (tablet) 
 *    windows that host sub-tasks like logging in and choosing player names, 
 *    tasks which take multiple steps and involve going backward and forward
 */
data class PanelContainerScreen<B : Any, T : Any>(
```